File: src/test/java/com/freya02/botcommands/test/commands/varargs/SlashVarargs.java
package com.freya02.botcommands.test.commands.varargs;

import com.freya02.botcommands.api.commands.CommandPath;
import com.freya02.botcommands.api.commands.application.ApplicationCommand;
import com.freya02.botcommands.api.commands.application.slash.ApplicationGeneratedValueSupplier;
import com.freya02.botcommands.api.commands.application.slash.GuildSlashEvent;
import com.freya02.botcommands.api.commands.application.slash.annotations.JDASlashCommand;
import com.freya02.botcommands.api.commands.application.slash.annotations.SlashOption;
import com.freya02.botcommands.api.commands.application.slash.annotations.VarArgs;
import com.freya02.botcommands.api.commands.application.slash.autocomplete.annotations.AutocompleteHandler;
import com.freya02.botcommands.api.parameters.ParameterType;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.events.interaction.command.CommandAutoCompleteInteractionEvent;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;
import java.util.List;

public class SlashVarargs extends ApplicationCommand {
	private static final String STR_AUTOCOMPLETE_NAME = "SlashVarargs: str";

	@Override
	public @NotNull ApplicationGeneratedValueSupplier getGeneratedValueSupplier(@Nullable Guild guild,
	                                                                            @Nullable String commandId, @NotNull CommandPath commandPath,
	                                                                            @NotNull String optionName, @NotNull ParameterType type) {
//		if (optionName.equals("number")) {
//			return e -> List.of(42L, 43L, 44L);
//		}

		return super.getGeneratedValueSupplier(guild, commandId, commandPath, optionName, type);
	}

	@JDASlashCommand(name = "varargs")
	public void run(GuildSlashEvent event,
	                @SlashOption(autocomplete = STR_AUTOCOMPLETE_NAME) String string,
	                @SlashOption(name = "number", description = "lol") @VarArgs(value = 3, numRequired = 2) List<Long> longs) {
		event.reply("longs: " + longs)
				.setEphemeral(true)
				.queue();
	}

	@AutocompleteHandler(name = STR_AUTOCOMPLETE_NAME)
	public Collection<String> onStrAutocomplete(CommandAutoCompleteInteractionEvent event, @SlashOption(name = "number") @VarArgs(3) List<Long> longs, @SlashOption String string) {
		return List.of(
				string + "_" + "str1" + "_" + longs.get(0),
				string + "_" + "str2" + "_" + longs.get(1),
				string + "_" + "str3" + "_" + longs.get(2),
				string + "_" + "str4"
		);
	}
}

File: src/test/kotlin/com/freya02/botcommands/test_kt/commands/slash/SlashVarargs.kt
package com.freya02.botcommands.test_kt.commands.slash

import com.freya02.botcommands.api.annotations.CommandMarker
import com.freya02.botcommands.api.commands.application.ApplicationCommand
import com.freya02.botcommands.api.commands.application.GlobalApplicationCommandManager
import com.freya02.botcommands.api.commands.application.annotations.AppDeclaration
import com.freya02.botcommands.api.commands.application.annotations.AppOption
import com.freya02.botcommands.api.commands.application.slash.GuildSlashEvent
import com.freya02.botcommands.api.commands.application.slash.annotations.JDASlashCommand
import com.freya02.botcommands.api.commands.application.slash.annotations.VarArgs
import dev.minn.jda.ktx.messages.reply_
import net.dv8tion.jda.api.events.interaction.command.CommandAutoCompleteInteractionEvent

@CommandMarker
class SlashVarargs : ApplicationCommand() {
    @JDASlashCommand(name = "varargs_annotated")
    fun onSlashVarargs(event: GuildSlashEvent, @AppOption @VarArgs(3, numRequired = 1) list: List<Int?>) {
        event.reply_("ok $list", ephemeral = true).queue()
    }

    @CommandMarker
    fun onVarargsAutocomplete(event: CommandAutoCompleteInteractionEvent, @AppOption list: List<Int?>): List<String> {
        return list.filterNotNull().map { it + 1 }.map { it.toString() }
    }

    @AppDeclaration
    fun declare(globalApplicationCommandManager: GlobalApplicationCommandManager) {
        globalApplicationCommandManager.slashCommand("varargs") {
            option("list") {
                varArgs = 3
                requiredVarArgs = 1

                autocomplete("SlashVarargs: onVarargsAutocomplete") {
                    showUserInput = false
                    function = ::onVarargsAutocomplete
                }
            }

            function = ::onSlashVarargs
        }
    }
}